File: Scala/src/com/yourtion/TinyWeb/TinyWeb.scala
package com.yourtion.TinyWeb

/**
  * Created by Yourtion on 9/5/16.
  */
class TinyWeb(controllers: Map[String, Controller],
              filters: List[(HttpRequest) => HttpRequest]) {

  def handleRequest(httpRequest: HttpRequest): Option[HttpResponse] = {
    val composedFilter = filters.reverse.reduceLeft(
      (composed, next) => composed compose next)
    val filteredRequest = composedFilter(httpRequest)
    val controllerOption = controllers.get(filteredRequest.path)
    controllerOption map { controller => controller.handleRequest(filteredRequest) }
  }
}

File: Java/src/com/yourtion/TinyWeb/TinyWeb.java
package com.yourtion.TinyWeb;

import java.util.List;
import java.util.Map;

/**
 * Created by Yourtion on 9/5/16.
 */
public class TinyWeb {
    private Map<String, Controller> controllers;
    private List<Filter> filters;

    public TinyWeb(Map<String, Controller> controllers, List<Filter> filters) {
        this.controllers = controllers;
        this.filters = filters;
    }

    public HttpResponse handleRequest(HttpRequest httpRequest) {

        HttpRequest currentRequest = httpRequest;
        for (Filter filter : filters) {
            currentRequest = filter.doFilter(currentRequest);
        }

        Controller controller = controllers.get(currentRequest.getPath());

        if (null == controller)
            return null;

        return controller.handleRequest(currentRequest);
    }
}
